plugin-server/src/main/ingestion-queues/session-recording/session-recordings-consumer-v3.ts

```diff
@@ -0,0 +1,450 @@
+    }
+
+    private async syncSessionsWithDisk(): Promise<void> {
+        // As we may get assigned and reassigned partitions, we want to make sure that we have all sessions loaded into memory
+        await Promise.all(
+            this.assignedTopicPartitions.map(async ({ partition }) => {
+                const keys = await readdir(path.join(this.rootDir, `${partition}`)).catch(() => {
+                    // This happens if there are no files on disk for that partition yet
+                    return []
+                })
+
+                // TODO: Below regex is a little crude. We should fix it
+                await Promise.all(
+                    keys
+                        .filter((x) => /\d+__[a-zA-Z0-9\-]+/.test(x))
+                        .map(async (key) => {
+                            // TODO: Ensure sessionId can only be a uuid
+                            const [teamId, sessionId] = key.split('__')
+
+                            if (!this.sessions[key]) {
+                                this.sessions[key] = await SessionManagerV3.create(this.config, this.objectStorage.s3, {
+                                    teamId: parseInt(teamId),
+                                    sessionId,
+                                    dir: this.dirForSession(partition, parseInt(teamId), sessionId),
+                                    partition,
+                                })
+                            }
+                        })
+                )
+            })
+        )
+    }
+
+    private async destroySession(key: string, sessionManager: SessionManagerV3): Promise<void> {
+        delete this.sessions[key]
+        await sessionManager.stop()
+    }
+
+    private setupHttpRoutes() {
+        expressApp.get('/api/projects/:projectId/session_recordings/:sessionId', async (req, res) => {
+            // TODO: Sanitize the projectId and sessionId as we are checking the filesystem
+
+            // validate that projectId is a number and sessionId is UUID like
+            const projectId = parseInt(req.params.projectId)
+            if (isNaN(projectId)) {
+                res.sendStatus(404)
+                return
+            }
+
+            const sessionId = req.params.sessionId
+            if (!/^[0-9a-f-]+$/.test(sessionId)) {
+                res.sendStatus(404)
+                return
+            }
+
+            status.info('🔁', 'session-replay-ingestion - fetching session', { projectId, sessionId })
+
+            // We don't know the partition upfront so we have to recursively check all partitions
+            const partitions = await readdir(this.rootDir).catch(() => [])
+
+            for (const partition of partitions) {
+                const sessionDir = this.dirForSession(parseInt(partition), projectId, sessionId)
+                const exists = await stat(sessionDir).catch(() => null)
+
+                if (!exists) {
+                    continue
+                }
```

Comment:
should we cache this? at busy times we'll be doing up to 128 file operations for every API request for data that changes very infrequently if ever

(might be getting ahead of myself)

Reply:
Step 1 - see if this is at all feasable using EFS cost wise.
Step 2 - optimize it.